Tutorial MemoryDB Example - Deploy MemoryDB Cluster Using the ACK MemoryDB Controller (#1249)
aws-controllers-k8s/community: docs/content/docs/tutorials/memorydb-example.md

---
title: "Deploy MemoryDB Cluster Using the ACK MemoryDB Controller"

> description: "Create managed MemoryDB Cluster from a Amazon Elastic Kubernetes Service (EKS) deployment."

Reviewer: I see you've taken this format from the RDS example, but I dislike using the term "deployment" here because it could be confused for a K8s `Deployment` type. Also I think these tutorials could be expanded to more than just EKS, so maybe even that's unnecessary. Here is my suggestion:

```suggestion
description: "Create a managed MemoryDB Cluster using the memorydb-controller"
```

(also nit: Please remove the trailing period)

Author: Updated the document.

> lead: "Create and use MemoryDB for Redis using Amazon Elastic Kubernetes Service (EKS)."

Reviewer: nit: Please remove the trailing period, and update to be the same as the description. Thanks

Author: Updated the document.

draft: false
menu:
  docs:
    parent: "tutorials"

> weight: 42

Reviewer 2: weight should be 44 to append it to existing tutorials

Author: Updated the document.

toc: true
---

The ACK service controller for Amazon MemoryDB for Redis lets you manage MemoryDB Cluster directly from Kubernetes.

This guide will show you how to create and connect to [MemoryDB for Redis](https://aws.amazon.com/memorydb/) through Kubernetes.

## Setup

> Although it is not necessary to use Amazon Elastic Kubernetes Service (Amazon EKS) with ACK, this guide assumes that you have access to an Amazon EKS cluster. If this is your first time creating an Amazon EKS cluster, see [Amazon EKS Setup](https://docs.aws.amazon.com/deep-learning-containers/latest/devguide/deep-learning-containers-eks-setup.html). For automated cluster creation using `eksctl`, see [Getting started with Amazon EKS - `eksctl`](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html) and create your cluster with Amazon EC2 Linux managed nodes.

Author: Updated the document.

> Although it is not necessary to use Amazon Elastic Kubernetes Service (Amazon EKS) with ACK, this guide assumes that you have access to an Amazon EKS cluster. If this is your first time creating an Amazon EKS cluster, see [Amazon EKS Setup](https://docs.aws.amazon.com/deep-learning-containers/latest/devguide/deep-learning-containers-eks-setup.html). For automated cluster creation using `eksctl`, see [Getting started with Amazon EKS - `eksctl`](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html) and create your cluster with Amazon EC2 Linux managed nodes.

Reviewer 2: nit: This is still not the updated Setup instructions.

Instruction from apigwv2 tutorial is
```
Although it is not necessary to use Amazon Elastic Kubernetes Service (Amazon EKS) with ACK, this guide assumes that you
have access to an Amazon EKS cluster. If this is your first time creating an Amazon EKS cluster, see
[Amazon EKS Setup][eks-setup]. For automated cluster creation using `eksctl`, see
[Getting started with Amazon EKS - `eksctl`][eksctl-guide].
```

Notice that "Amazon EC2 Linux Managed nodes" part is not present and the lengthy urls are replaced by references in the end of the document for better readability.

### Prerequisites

This guide assumes that you have:

- Created an EKS cluster with Kubernetes version 1.18 or higher.
- Setup the [Amazon VPC Container Network Interface (CNI) plugin for Kubernetes](https://docs.aws.amazon.com/eks/latest/userguide/managing-vpc-cni.html) for the EKS Cluster.
- AWS IAM permissions to create roles and attach policies to roles.
- Installed the following tools on the client machine used to access your Kubernetes cluster:
  - [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv1.html) - A command line tool for interacting with AWS services.
  - [kubectl](https://docs.aws.amazon.com/eks/latest/userguide/install-kubectl.html) - A command line tool for working with Kubernetes clusters.
  - [eksctl](https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html) - A command line tool for working with EKS clusters.

>   - [Helm 3.7+](https://helm.sh/docs/intro/install/) - A tool for installing and managing Kubernetes applications.

Reviewer: I think this would be the best time to start recommending Helm 3.8. It has native support for OCI registries like we are using. We don't need to do the whole `helm pull` + `untar` + `helm install`. We can just directly use `helm install` and pass it the OCI image URI!

Reviewer 2: I think `untar` is currently needed because we use local files to delete the crds in cleanup [step](https://github.com/aws-controllers-k8s/community/blob/main/docs/content/docs/user-docs/cleanup.md#delete-all-crds).

Maybe we can delete crds by name instead of files and then get rid of untar

Author: I have kept the Helm3.7 specific instructions for now.
I think it can be updated later across all the examples files.

### Install the ACK service controller for MemoryDB

You can deploy the ACK service controller for Amazon MemoryDB using the [memorydb-chart Helm chart](https://gallery.ecr.aws/aws-controllers-k8s/memorydb-chart). You can download it to your workspace using the following command:

```bash
helm pull oci://public.ecr.aws/aws-controllers-k8s/memorydb-chart --version=v0.0.1
````

You will need to decompress and extract the Helm chart. You can do so with the following command:

```bash
tar xzvf memorydb-chart-v0.0.1.tgz
```

You can now use the Helm chart to deploy the ACK service controller for Amazon MemoryDB to your EKS cluster. At a minimum, you need to specify the AWS Region to execute the MemoryDB API calls.

For example, to specify that the MemoryDB API calls go to the `us-east-1` region, you can deploy the service controller with the following command:

```bash
helm install memorydb-chart --generate-name --set=aws.region=us-east-1

> ```

Reviewer: Please update the code an documentation to reflect the latest Helm 3.8 CLI options:

```
# Log into the chart repository
aws ecr-public get-login-password --region us-east-1 | helm registry login -u AWS --password-stdin public.ecr.aws

# Install the memorydb-chart Helm chart
helm install \
    --generate-name \
    --namespace ack-system \
    --set aws.region=us-east-1 \
    oci://public.ecr.aws/aws-controllers-k8s/memorydb-chart \
    --version v0.0.1
```

Author: I have kept the Helm3.7 specific instructions for now.
I think it can be updated later across all the examples files.

For a full list of available values to the Helm chart, please [review the values.yaml file](https://github.com/aws-controllers-k8s/memorydb-controller/blob/main/helm/values.yaml).

### Configure IAM permissions

Once the service controller is deployed, you will need to [configure the IAM permissions][irsa-permissions] for the controller to query the MemoryDB API. For full details, please review the AWS Controllers for Kubernetes documentation for [how to configure the IAM permissions][irsa-permissions]. If you follow the examples in the documentation, use the value of `memorydb` for `SERVICE`.

## Deploy MemoryDB Cluster Instances

You can deploy MemoryDB Clusters using the `Cluster` custom resource. The examples below show how to deploy it from your Kubernetes environment. For a full list of options available in the `Cluster` custom resource definition, you can use `kubectl explain cluster`.

The examples below use the `db.t4g.small` node type. Please review the [MemoryDB node types](https://docs.aws.amazon.com/memorydb/latest/devguide/nodes.supportedtypes.html) to select the most appropriate one for your workload.

### MemoryDB Cluster

> To create a MemoryDB Cluster, create a `Cluster` custom resource. The example below shows how to provision a MemoryDB Cluster:

Author: Updated the document.

#### Create MemoryDB Cluster in default VPC

> Following yaml creates MemoryDB Cluster using default VPC subnets and security group.

Reviewer: nit: grammar
```suggestion
The following YAML creates a MemoryDB Cluster using the default VPC subnets and security group
```

Author: Updated the document.

```bash

> MEMORYDB_CLUSTER_NAME="<your cluster name>"

Author: Updated the document.

cat <<EOF > memorydb-cluster.yaml
apiVersion: memorydb.services.k8s.aws/v1alpha1
kind: Cluster
metadata:
  name: "${MEMORYDB_CLUSTER_NAME}"
spec:
  name: "${MEMORYDB_CLUSTER_NAME}"
  nodeType: db.t4g.small
  aclName: open-access
EOF

kubectl apply -f memorydb-cluster.yaml
```

#### Create MemoryDB Cluster in specific VPC subnets and security groups

> To provision MemoryDB Cluster using specific Subnets from a VPC, create MemoryDB `SubnectGroup` custom resource first and then specify it and the security group name in `Cluster` specification:

Reviewer: ```suggestion
To provision a MemoryDB Cluster using specific subnets from a VPC, create a MemoryDB `SubnectGroup` custom resource first and then specify it, and a security group name, in the `Cluster` specification.
```

Author: Updated the document.

> To provision MemoryDB Cluster using specific Subnets from a VPC, create MemoryDB `SubnectGroup` custom resource first and then specify it and the security group name in `Cluster` specification:

Author: Updated the document.

##### Create MemoryDB Subnet Group

> The following example uses the VPC ID of the EKS Cluster. You may specify any other VPC id as well.

Reviewer: ```suggestion
The following example uses the VPC ID of the EKS Cluster. You may specify any other VPC ID by modifying the list of `subnetIDs` in the specification.
```

Author: Updated the document.

```bash
EKS_CLUSTER_NAME="<your EKS cluster name>"

> VPC_ID=$(aws eks describe-cluster --name $EKS_CLUSTER_NAME --query cluster.resourcesVpcConfig.vpcId)

Reviewer: Add region?

Author: Added region to the commands.

SUBNET_IDS=$(aws ec2 describe-subnets \
  --filters "Name=vpc-id,Values=${VPC_ID}" \
  --query 'Subnets[*].SubnetId' \
  --output text
)

> MEMORYDB_SUBNETGROUP_NAME="<your subnet group name>"

Author: Updated the document.

cat <<EOF > memorydb-subnetgroup.yaml
apiVersion: memorydb.services.k8s.aws/v1alpha1
kind: SubnetGroup
metadata:
  name: "${MEMORYDB_SUBNETGROUP_NAME}"
spec:
  name: "${MEMORYDB_SUBNETGROUP_NAME}"
  description: "MemoryDB cluster subnet group"
  subnetIDs:
$(printf "    - %s\n" ${SUBNET_IDS})

EOF

kubectl apply -f memorydb-subnetgroup.yaml
kubectl describe subnetgroup "${MEMORYDB_SUBNETGROUP_NAME}"
```
##### Create MemoryDB Cluster

> The following example uses the MemoryDB Subnet Group created above. It uses EKS Cluster Security group. You may use any other VPC security group as well.

Reviewer: ```suggestion
The following example uses the MemoryDB Subnet Group created above. It uses the provisioning EKS Cluster's security group. You may specify any other security group by modifying the list of `securityGroupIDs` in the specification.
```

Author: Updated the document.

```shell

> EKS_CLUSTER_NAME="<your EKS cluster name>"

Reviewer: This can be re-used from above? Maybe we should just use a `export EKS_CLUSTER_NAME` at the top, after asking the user to create the cluster?

Author: I preferred keeping the relevant variables close to the commands.

MEMORYDB_CLUSTER_NAME="<your MemoryDB cluster name>"

> SECURITY_GROUP_ID=$(aws eks describe-cluster --name $EKS_CLUSTER_NAME --query cluster.resourcesVpcConfig.clusterSecurityGroupId)

Reviewer: Add region

Author: Added region to the commands.

cat <<EOF > memorydb-cluster.yaml
apiVersion: memorydb.services.k8s.aws/v1alpha1
kind: Cluster
metadata:
  name: "${MEMORYDB_CLUSTER_NAME}"
spec:
  name: "${MEMORYDB_CLUSTER_NAME}"
  nodeType: db.t4g.small
  aclName: open-access
  securityGroupIDs:

>     - SECURITY_GROUP_ID

Reviewer: Oops

Author: Thank you, updated to use variable syntax.

  subnetGroupName: ${MEMORYDB_SUBNETGROUP_NAME}
EOF

kubectl apply -f memorydb-cluster.yaml
kubectl describe cluster "${MEMORYDB_CLUSTER_NAME}"
```

You can track the status of the provisioned database using `kubectl describe` on the `Cluster` custom resource:

```bash

> kubectl describe cluster "${MEMORYDB_CLUSTER_NAME}"

Author: Updated the document and added example output.

```

> When the `Cluster Status` says `Available`, you can connect to the database instance.

Reviewer 3: nit: `available`

Author: Updated the document.

## Connect to MemoryDB Cluster

>To connect to the MemoryDB Cluster from a Pod running inside Kubernetes cluster:
> * Ensure that the [Amazon VPC Container Network Interface (CNI) plugin for Kubernetes](https://docs.aws.amazon.com/eks/latest/userguide/managing-vpc-cni.html) has been setup for the EKS Cluster.
> * Review [access patterns for accessing a MemoryDB Cluster in an Amazon VPC](https://docs.aws.amazon.com/memorydb/latest/devguide/memorydb-vpc-accessing.html) to confirm that MemoryDB cluster is configured to allow connection from the Pod.

The `Cluster` status contains the information for connecting to a MemoryDB Cluster. The host information can be found in `status.ClusterEndpoint.Address` and the port information can be found in `status.ClusterEndpoint.Port`. For example, you can get the connection information for a `Cluster` created in one of the previous examples using the following commands:

```bash
kubectl get cluster "${MEMORYDB_CLUSTER_NAME}" -o jsonpath='{.status.ClusterEndpoint.Address}'

> kubectl get cluster "${MEMORYDB_CLUSTER_NAME}" -o jsonpath='{.status.ClusterEndpoint.Port}'

Reviewer 3: ``` 
kubectl get cluster "${MEMORYDB_CLUSTER_NAME}" -o jsonpath='{.status.clusterEndpoint.address}'
kubectl get cluster "${MEMORYDB_CLUSTER_NAME}" -o jsonpath='{.status.clusterEndpoint.port}'
```

Need to fix case in above paragraph as well

Author: Thanks, updated the document.

```

You can extract this information and make it available to your Pods using a [`FieldExport`][field-export] resource. For example, to get the connection information from MemoryDB Cluster created the above example, you can use the following example:

```shell

> MEMORYDB_CLUSTER_NAME="<your cluster name>"

Reviewer: Like `EKS_CLUSTER_NAME`, lets re-use this value from above

Author: I preferred keeping the relevant variables close to the commands.

            valueFrom:
              configMapKeyRef:
                name: ${MEMORYDB_CLUSTER_CONN_CM}
                key: "${MEMORYDB_CLUSTER_NAME}-host"
          - name: MEMORYDB_CLUSTER_PORT
            valueFrom:
              configMapKeyRef:
                name: ${MEMORYDB_CLUSTER_CONN_CM}
                key: "${MEMORYDB_CLUSTER_NAME}-port"
EOF

kubectl apply -f game_leaderboard.yaml
```

## Next steps

You can learn more about each of the ACK service controller for MemoryDB custom resources by using `kubectl explain` on the API resources. These include:

* `cluster`
* `user`
* `acl`
* `parametergroup`
* `subnetgroup`

The ACK service controller for Amazon MemoryDB is based on the [Amazon MemoryDB API](https://docs.aws.amazon.com/memorydb/latest/APIReference/Welcome.html). To get a full understanding of how all of the APIs work, please review the [Amazon MemoryDB API documentation](https://docs.aws.amazon.com/memorydb/latest/APIReference/Welcome.html).

You can learn more about [how to use Amazon MemoryDB](https://docs.aws.amazon.com/memorydb/index.html) through the [documentation](https://docs.aws.amazon.com/memorydb/index.html).

### Cleanup

> You can deprovision your MemoryDB cluster using `kubectl delete` command.

Reviewer: And the rest of the resources?

Author: Updated the document to delete the example resources that the tutorial creates.
